libxfce4ui/xfce-gtk-extensions.c

```diff
@@ -49,6 +49,464 @@
+      g_warning ("xfce_gtk_menu_item_new_from_action_entry: Unknown item_type");
+      return NULL;
+    }
+}
+
+
+
+/**
+ * xfce_gtk_toggle_menu_item_new_from_action_entry:
+ * @action_entry : Label to use for the #GtkCheckMenuItem
+ * @callback_param : optional callback parameter, or NULL.
+ * @menu_to_append_item : #GtkMenuShell on which the item should be appended, or NULL
+ *
+ * Method to create a toggle menu item from the passed action entry
+ *
+ * Return value: (transfer full): A new #GtkMenuItem or NULL
+ *
+ * Since: 4.16
+ **/
+GtkWidget*
+xfce_gtk_toggle_menu_item_new_from_action_entry (const XfceGtkActionEntry *action_entry,
+                                                 GObject                  *callback_param,
+                                                 gboolean                  active,
+                                                 GtkMenuShell             *menu_to_append_item)
+{
+  if (action_entry->menu_item_type == XFCE_GTK_CHECK_MENU_ITEM)
+    {
+      return xfce_gtk_check_menu_item_new (action_entry->menu_item_label_text, action_entry->menu_item_tooltip_text,
+                                           action_entry->accel_path, action_entry->callback,
+                                           callback_param, active, menu_to_append_item);
+    }
+  else if (action_entry->menu_item_type == XFCE_GTK_RADIO_MENU_ITEM)
+    {
+      return xfce_gtk_radio_menu_item_new (action_entry->menu_item_label_text, action_entry->menu_item_tooltip_text,
+                                           action_entry->accel_path, action_entry->callback,
+                                           callback_param, active, menu_to_append_item);
+    }
+  else
+    {
+      g_warning ("xfce_gtk_toggle_menu_item_new_from_action_entry: Unknown item_type");
+      return NULL;
+    }
+}
+
+
+
+/**
+ * xfce_gtk_tool_button_new_from_action_entry:
+ * @action_entry : Label to use for the #GtkTolButton
+ * @callback_param : optional callback parameter, or NULL.
+ * @toolbar_to_append_item : #GtkToolBar on which the item should be appended
+ *
+ * Method to create a toolbar button from the passed action entry
+ *
+ * Return value: (transfer full): A new #GtkTolButton
+ *
+ * Since: 4.16
+ **/
+GtkWidget*
+xfce_gtk_tool_button_new_from_action_entry (const XfceGtkActionEntry *action_entry,
+                                            GObject                  *callback_param,
+                                            GtkToolbar               *toolbar_to_append_item)
+{
+  GtkToolItem *tool_item;
+  GtkWidget   *image;
+
+  image = gtk_image_new_from_icon_name (action_entry->menu_item_icon_name, GTK_ICON_SIZE_LARGE_TOOLBAR);
+  tool_item = gtk_tool_button_new (image, action_entry->menu_item_label_text);
+  g_signal_connect_swapped (G_OBJECT (tool_item), "clicked", action_entry->callback, callback_param);
+  gtk_widget_set_tooltip_text (GTK_WIDGET (tool_item), action_entry->menu_item_tooltip_text);
+  gtk_widget_show (GTK_WIDGET (tool_item));
```

Comment:
I'm not sure about calling gtk_widget_show() from library functions... What if the caller wants to show/hide the UI elements separately or on demand?
Do GTK library functions call gtk_widget_show()?

Reply:
No, GTK library functions dont call `gtk_widget_show` automatically.
I did so in order to save some(many) lines of code, since in the most use-cases the widget is added to be shown.
You still can call `gtk_widget_hide` after calling this method in order to hide the widget.

Reply:
Yeah, I know it saves lines of code on the client side, but, on the other hand, I'd rather be consistent with what GTK does. It's probably no big deal for the client to call `gtk_widget_show()` since it already does that for numerous GTK widgets.

Reply:
Hmm, actually I dont want to be consistant to GTK :) 
I already added all the nullable arguments for the same reason: To save all these endlessly repeading lines of code.
I have to admit that the arguments are more visible, while gtk_widget_show is done silently ... on the other hand, it's a custom implementation, so I guess differences to GTK can be expected.

Removing gtk_widget_show( ) in all methods here would require ~70 additional lines of code in thunar.

@andreldm, @ochosi : What would be your preference on this ?

Reply:
Why so many lines of code in thunar? Does it use 'gtk_widget_show_all()` or calls `gtk_widget_show()` for each element individually?

Reply:
True, I did not think about `gtk_widget_show_all()` ... so probably only a handfull of lines .... ok, I'll check if it does the job, and remove gtk_widget_show() from here.

Reply:
Thanks!

Reply:
pushed.
Thank you for pointing at it !